src/cli/serve/serve.js

```diff
@@ -78,25 +78,44 @@ const configPathCollector = pathCollector();
 const pluginPathCollector = pathCollector();
 
 /**
- * @param {string} name
- * @param {string[]} configs
- * @param {'push' | 'unshift'} method
+ * @param {string} name The config file name
+ * @returns {boolean} Whether the file exists
  */
-function maybeAddConfig(name, configs, method) {
+function configFileExists(name) {
   const path = resolve(getConfigDirectory(), name);
   try {
-    if (statSync(path).isFile()) {
-      configs[method](path);
-    }
+    return statSync(path).isFile();
   } catch (err) {
     if (err.code === 'ENOENT') {
-      return;
+      return false;
     }
 
     throw err;
   }
 }
 
+/**
+ * @returns {boolean} Whether the distribution can run in Serverless mode
+ */
+function isServerlessCapableDistribution() {
+  // For now, checking if the `serverless.yml` config file exists should be enough
+  // We could also check the following as well, but I don't think it's necessary:
+  // VALID_SERVERLESS_PROJECT_MODE.some((projectType) => configFileExists(`serverless.${projectType}.yml`))
+  return configFileExists('serverless.yml');
+}
+
+/**
+ * @param {string} name
+ * @param {string[]} configs
+ * @param {'push' | 'unshift'} method
+ */
+function maybeAddConfig(name, configs, method) {
```

Comment:
nitnit: I know this is a pre-existing function, but I don't find the API very readable

```ts
    // we "unshift" .serverless. config so that it only overrides defaults
    if (serverlessMode) {
      maybeAddConfig(`serverless.yml`, configs, 'push');
      maybeAddConfig(`serverless.${serverlessMode}.yml`, configs, 'unshift');
    }
```

IMO something like this would be clearer:

```ts
    if (serverlessMode) {
      configs.push(resolveConfig(`serverless.yml`));
      // we "unshift" .serverless. config so that it only overrides defaults
      configs.unshift(resolveConfig(`serverless.${serverlessMode}.yml`));
    }
    ...
    configs = configs.filter(Boolean);
```
where `resolveConfig: (config: string) => undefined | string`

Reply:
++ It looks much cleaner your way!

I've created https://github.com/elastic/kibana/issues/155137 to tackle it as an entity change on its own. I'll address it as soon as this one is merged